rclcpp/include/rclcpp/service.hpp

```diff
@@ -100,25 +105,50 @@ class Service : public ServiceBase
     using rosidl_generator_cpp::get_service_type_support_handle;
     auto service_type_support_handle = get_service_type_support_handle<ServiceT>();
 
+    // rcl does the static memory allocation here
+    service_handle_ = new rcl_service_t;
+    *service_handle_ = rcl_get_zero_initialized_service();
+
     if (rcl_service_init(
-        &service_handle_, node_handle.get(), service_type_support_handle, service_name.c_str(),
+        service_handle_, node_handle.get(), service_type_support_handle, service_name.c_str(),
         &service_options) != RCL_RET_OK)
     {
-      throw std::runtime_error(
-              std::string("could not create service: ") +
+      throw std::runtime_error(std::string("could not create service: ") +
               rcl_get_error_string_safe());
     }
   }
 
+  Service(
+    std::shared_ptr<rcl_node_t> node_handle,
+    rcl_service_t * service_handle,
+    AnyServiceCallback<ServiceT> any_callback)
+  : ServiceBase(node_handle),
+    any_callback_(any_callback)
+  {
+    // check if service handle was initialized
+    // see: https://github.com/ros2/rcl/issues/81
+    if (service_handle->impl == NULL) {
+      throw std::runtime_error(
+              std::string("rcl_service_t in constructor argument ") +
+              "has to be initialized beforehand");
```

Comment:
This must be a copy-paste issue, normally that pattern is used like this:

```c++
throw std::runtime_error(
  std::string("there was an error while trying to do something: ") +
  rcl_get_error_string_safe());
```

Where the first item needs to be a `std::string` so that the `+` operator works with the `const char *` returned by the error function.